Bug fixes

dhull opened the pull request:
This pull request contains three bug fixes.
- There was a pattern match error in one of the `do_work` heads that would prevent it from matching cast or info requests. This would cause the `gen_server_pool` to crash on these requests instead of forwarding the request to the proxy.
- When a request was dropped from the work queue because it was full, the request_dropped response was sent to the process that sent the new request that caused the old request to be dropped instead of to the process that sent the old request that was being dropped.
- When a request was timed out because the it was removed from the work queue after its request_max_wait, the worker that was taken from the worker pool to handle the request lost because it was not returned to the worker pool. In addition, there was no stats accounting for these requests; they are now counted as dropped tasks.

There's also some code cleanup.

Finally, gen_server_pool has a problem that it doesn't notice that a worker is no longer alive until it attempts to use it to handle a request. This is not a huge issue, but it does mean that the pool can get filled with worker entries for workers that no longer exist. In this pull request I have added some logic in the idle-check code to clean up these workers and to kill workers that have exceeded their lifetime. I'm not really happy with this solution, but it seems like an improvement over the existing functionality. I will work on this more in the future.

Commit:
Fix do_work pattern match for cast or info requests.

The function head wanted to match the request's From pid, but this match
would fail for non-$call requests.  This bug was introduced when
request_max_wait support was added.

Commit:
Move handle_request function to be near do_work, which it calls.

Commit:
Send request_dropped error reply directly from do_work.

When a call is dropped because there are no workers available and the
queue is full, send the request_dropped reply directly from do_work.
This fixes a bug where the request that is dropped is the oldest request
in the queue, but previously the request_dropped reply is sent to the
request that was just received.

This is also more consistent with how the request_timeout reply is sent
and will allow some refactoring in the future.

Commit:
Since do_work no longer returns errors, have it return new state directly.

Commit:
Use the queue module instead of a home-grown FIFO.

Commit:
When handling a request timeout don't lose worker and try next request.

When a request is timed out because it is too old, the worker that was
taken from the available worker stack to handle the request was
forgotten.

In addition, do_work just returned even though there might be other
requests on the queue; now it tries the next request.

Finally, there was not accounting for these timeout-out requests.  They
are now counted as dropped tasks.

Commit:
Add .rebar directory to .gitignore.

Commit:
No need to have schedule_idle_check return state.

Commit:
Use lists:keyfind BIF instead of proplists:is_defined for speed.

Commit:
Check entire worker pool for dead and too-old workers during idle check.

Commit:
Add test for gen_server_pool call error_dropped.

Commit:
Makefile: Add "dialyzer" rule, and make "check" be a synonym for "test".

Commit:
Add test for gen_server_pool cast.

hirotnk commented on src/gen_server_pool.erl:
```diff
@@ -347,14 +369,14 @@ do_work( State = #state{ proxy_ref = ProxyRef,
-                                num_queued_tasks=NumTasks-1}
-              }
-           end
+              State#state{ available = Workers,
```
Is it a bad idea to make a recursive call to 'do_work' here ?

dhull replied:
By "here" you mean line 347/369? I guess that there might be a problem if the worker that we killed was the last one in the pool. In this case we kill it and start a new one, but the new one isn't going to be in the pool until "available" is 'call'ed and we can't handle that until we return from this 'call', so in that case we would start another new worker. So I think its ok to call `do_work` if the Workers list is not empty, otherwise we should wait for the new worker we initialized to become available.

It does seem like this would happen only once, when there's only one worker in the pool and it's too old, we end up starting two new workers instead of one.

dhull replied:
Oh, I see, by "here" you mean line 372 in the new code. I think it is a bad idea to call `do_work` there, because if we do we can start up an excessive number of new workers. Imagine this sequence:
1. the queue is empty but all the workers are busy. but then N requests come in, and we're below the max number of workers so we start a new worker for each of them.
2. the first new worker calls available, is put in the pool, and then do_work is called.
3. we get to this `do_work` head, and pass the oldest request in the queue to this worker.
4. if we call `do_work` again, we will start another new worker, which we don't want to do, because we already started a new worker for this request when it was queued.

hirotnk replied:
>  4 . if we call do_work again, we will start another new worker, which we don't want to do, because we already started a new worker for this request when it was queued.

If we have N requests in the queue, don't we still check the pool size with max pool size as in `case PoolSize < MaxSize of` ?

hirotnk commented on src/gen_server_pool.erl:
```diff
@@ -285,54 +271,90 @@ emit_stats( #state{ prog_id = ProgId, pool_id = PoolId } = S ) ->
-            }
-          };
+          %% Queue too big, so drop the oldest request.
+          { { value, #request{ call_args = CallArgs } }, RequestsOut } = queue:out( Requests ),
```
It looks still returning the message 'request_dropped' to the process who sent the oldest request instead of the process who sent the latest request. I guess it doest not matter much though ?

dhull replied:
Yes, that's what you want to do. You have a max queue length of N and N+1 requests that want to be served. You drop the oldest, because that's the one that's least likely to be served before its deadline expires.

hirotnk commented on src/gen_server_pool.erl:
```diff
@@ -347,14 +369,14 @@ do_work( State = #state{ proxy_ref = ProxyRef,
               do_work( State#state { available = Workers } );
```
On the line 368, 'assure_min_pool_size/1' can return error. I think it's better to log the error somewhere when it happens.

hirotnk commented on src/gen_server_pool.erl:
```diff
@@ -414,35 +415,85 @@ check_idle_timeouts( #state{ proxy_ref = ProxyRef,
                              available = Available,
-                             min_size = MinSize } = S ) ->
+                             min_size = MinSize,
+                             max_worker_age = MaxWorkerAge } = S ) ->
```
It's not only this one, but it's hard to figure what's the unit of this field by looking at the field name.

dhull replied:
> It's not only this one, but it's hard to figure what's the unit of this field by looking at the field name.

I agree 100%. I plan to address that in a future commit.

djnym commented:
It all sounds good to me, the code seems as reasonable as it can be.  Is there any reason to leave the available workers as a stack and not use a queue for that as well?  It might make it easier as you won't ever have idle workers?

dhull commented:
> Is there any reason to leave the available workers as a stack and not use a queue for that as well? It might make it easier as you won't ever have idle workers?

I think it's best to use a stack instead of a queue because we want to leave idle workers idle so that if we spin up extra workers to handle a burst of requests we allow those workers to go idle and be killed by the idle timeout even if they don't have a lifetime.

Commit:
Major do_work refactoring.

Commit:
Use term "request" instead "task" for consistency.

Commit:
Fix use of proplists calls on available list.

The available list is a list of worker records, not a proplist, so the
proplist functions are incorrect.

Commit:
Rename min_size to min_pool_size and max_size to max_pool_size.

Commit:
Validate option values in parse_opts.

Commit:
Rename options and fields for clarity; consistently use microseconds for relative times.

Commit:
Add gen_server_pool_starter module to start workers asynchronously.

dhull commented:
I noticed that `gen_server_pool_proxy:handle_call` has the comment:
```
  % [...], since if an embedded gen_server is using gen_server:reply/2
  % and thus using the noreply forms, we don't want to make the resource
  % available.
```
And if the proxied gen_server returns a `noreply` response it indeed does not call `gen_server_pool:available` on the worker.

This doesn't make sense to me. The `handle_call` has returned and the worker is going to go back into the `gen_server`'s receive loop. Why would we *not* want to return the worker to the pool? Otherwise, it's still alive but not able to receive requests.

Commit:
Move test suite to test directory.

hirotnk commented:
I think it is calling `gen_server_pool:available` in `handle_info` in that case.

dhull commented:
> I think it is calling gen_server_pool:available in handle_info in that case.

You're right. I was confused, but Anthony was just here and I understand now.

Commit:
Add synchronous gen_server_pool:stop/1 interface.

It previously wasn't possible to shut down a gen_server_pool.

Commit:
Move tests to test directory and improve them.

Commit:
Use a monitor to track when workers die.

Use this to track how many workers are alive and to remove workers from
the available list as they die.  Potentially this could be used to
eliminate many of the calls to gen_server_pool_starter:worker_count.

Commit:
Enhance worker lifetime handling.

Previously, max_worker_age_ms could only be specified as a lifetime in
milliseconds or 'infinity'.  Now it can also be specified as
{LifetimeMS, JitterMS}, and {Module, Function, Args}.  Adding jitter to
the worker lifetimes can be useful when the worker's initialization is
expensive and you want to avoid having lots of workers restart at the
same time.

To support each workers having a different lifetime, its end time is now
calculated when the worker is started and, instead of the start time,
this end time is carried around with the worker's state.

Commit:
Add a test of max_worker_idle_ms.

dhull commented:
I just pushed another batch of commits. The first ("Fix use of proplists calls on available list.") fixes another existing bug in gen_server_pool. The rest:
* Attempt to simplify the logic (although it ends up being more code, it's less-tricky code).
* Add a new gen_server_pool_starter module to perform asynchronous starting of new workers. This should reduce gen_server_pool's latency.
* Add a `gen_server_pool:stop` function to allow the pool to be shut down.
* Use monitors to track workers so that they are removed from the available pool when they die.
* Enhanced the worker lifetime handling to allow some randomness to be added. This should allow the functionally-equivalent code in MOPS to be removed.

djnym commented on src/gen_server_pool_internal.hrl:
```diff
@@ -0,0 +1,6 @@
+-ifndef(_gen_server_pool_internal_included).
+-define(_gen_server_pool_internal_included, true).
+
+-define(GSP_FLAG_INITIAL, 1).
```
So I understand what this is doing, but I think if you go down this route, you should also define the non initial flag, as I see you just have bare '0's in other calls to gen_server_pool:available/5?

dhull replied:
This is a C-ism. I initially used an atom here, but I couldn't decide what the opposite of "initial" was, so I decided to sidestep the problem with a representation where I didn't have to. So, there isn't really a "non-initial" flag, but I could define GCP_FLAG_NONE as 0 if you would prefer.

djnym replied:
Yeah, that at least would hide the implementation detail and sounds good.  An atom or boolean would also work.  I just don't like a define in one place and a random 0 somewhere else.

djnym commented on src/gen_server_pool.erl:
```diff
@@ -189,8 +222,9 @@ handle_info( Info, State ) ->
-  terminate_pool( Reason, State ),
+terminate( Reason, #state{ sup_pid = SupPid, starter_ref = StarterRef } ) ->
+  exit( SupPid, Reason ),
+  gen_server_pool_starter:stop( StarterRef ),
```
I still think this should be put into some sort of supervision tree, because this just seems ugly.  At the very least maybe a FIXME comment to that effect?

hirotnk commented on src/gen_server_pool.erl:
```diff
@@ -189,8 +222,9 @@ handle_info( Info, State ) ->
-terminate( Reason, State ) ->
-  terminate_pool( Reason, State ),
+terminate( Reason, #state{ sup_pid = SupPid, starter_ref = StarterRef } ) ->
+  exit( SupPid, Reason ),
```
Why it's killing `SupPid` when it's already linked ?

dhull replied:
Yes, this seems unnecessary.

hirotnk commented on src/gen_server_pool_starter.erl:
```diff
@@ -0,0 +1,141 @@
+
+
+stop( { SelfPid, _ } ) ->
+  exit( SelfPid, normal ).
```
Why not use `gen_server:stop/1 or 3` or `gen_server:call/2 or 3` ?

dhull replied:
`gen_server:stop` is not a documented interface in R16, and so I was not aware of it, but it does look equivalent. I'll make that change.

dhull replied:
Actually, `gen_server:stop/1` does not exist in Erlang R16, so I'm going to keep this the way it is. In Erlang 18, `gen_server:stop(Process)` calls `stop(Process, normal, infinity)`, so my call is equivalent.

hirotnk replied:
It does not look equivalent since it's using message to make it synchronous termination at first, but I have no objection to keep it that way.

dhull replied:
Are you looking at something different than I am (this is OTP 18.2.1)?
In `otp/lib/stdlib/src/gen_server.erl`:
```
stop(Name) ->
    gen:stop(Name).
```
In `otp/lib/stdlib/src/gen.erl`:
```
stop(Process) ->
    stop(Process, normal, infinity).
```

dhull replied:
So, I'm blind, and read that last `stop` as `exit` or something, but Yoshi was very patient in pointing out my misunderstanding. I will have a fix for this shortly.

dhull commented:
I just pushed another set up updates that I believe address all of the outstanding comments.

1. I added a `GSP_FLAGS_NONE` define to make the difference between the initial and subsequent calls to `gen_server_pool:available` more clear.
2. I added a "FIXME" comment about the lack of a proper supervision tree.
3. I got rid of an `exit` call that was unnecessary because the process being signaled was already linked.
4. I didn't get rid of a second `exit` call because the proposed replacement with `gen_server:stop` doesn't work in Erlang R15.

djnym commented:
So I think it looks good, I think the only thing is to bump the version in the .app.src file and add a ChangeLog entry and then we can merge and release.  Is this worth a major version bump?

Commit:
Update ChangeLog and bump version to 2.4.0.

hirotnk commented:
Looks good.
